src/annoylib.h

```diff
@@ -430,6 +428,90 @@ struct Base {
   }
 };
 
+
+struct Cosine : Base {
+  template<typename S, typename T>
+  struct Node {
+    /*
+     * We store a binary tree where each node has two things
+     * - A vector associated with it
+     * - Two children
+     * All nodes occupy the same amount of memory
+     * All nodes with n_descendants == 1 are leaf nodes.
+     * A memory optimization is that for nodes with 2 <= n_descendants <= K,
+     * we skip the vector. Instead we store a list of all descendants. K is
+     * determined by the number of items that fits in the space of the vector.
+     * For nodes with n_descendants == 1 the vector is a data point.
+     * For nodes with n_descendants > K the vector is the normal of the split plane.
+     * Note that we can't really do sizeof(node<T>) because we cheat and allocate
+     * more memory to be able to fit the vector outside
+     */
+    S n_descendants;
+    union {
+      S children[2]; // Will possibly store more than 2
+      T norm;
+    };
+    T v[V_ARRAY_SIZE];
+  };
+  template<typename S, typename T>
+  static inline T distance(const Node<S, T>* x, const Node<S, T>* y, int f) {
+    // want to calculate:
+    // 1 - ((a.b) / |a|*|b|)
+    T mx = get_norm(x->v, f);
+    T my = get_norm(y->v, f);
+    T d_xy = dot(x->v, y->v, f);
+    return 1 - (d_xy / (mx*my));
```

Comment:
Just to double check
The range of `(d_xy / (mx*my))` is between -1 to 1, hence `1 - (d_xy / (mx*my))` goes from 2 to 0.
The less similar (similarity = -1) becomes distance 2. Seems ok.

Reply:
However `(mx*my)` can be zero causing a division by 0, in which case you should return `1-0 = 1`